src/stream/src/executor/over_window/mod.rs

```diff
@@ -0,0 +1,692 @@
+                input_pk_indices: input_info.pk_indices,
+                partition_key_indices: args.partition_key_indices,
+                order_key_index: args.order_key_index,
+                watermark_epoch: args.watermark_epoch,
+            },
+        }
+    }
+
+    async fn ensure_key_in_cache(
+        this: &mut ExecutorInner<S>,
+        cache: &mut PartitionCache,
+        partition_key: impl Row,
+        encoded_partition_key: &MemcmpEncoded,
+    ) -> StreamExecutorResult<()> {
+        if cache.contains(encoded_partition_key) {
+            return Ok(());
+        }
+
+        let mut partition = Partition::new(&this.calls);
+
+        // Recover states from state table.
+        let table_iter = this
+            .state_table
+            .iter_with_pk_prefix(partition_key, PrefetchOptions::new_for_exhaust_iter())
+            .await?;
+
+        #[for_await]
+        for row in table_iter {
+            let row: OwnedRow = row?;
+            let order_key = row
+                .datum_at(
+                    this.col_mapping
+                        .upstream_to_state_table(this.order_key_index)
+                        .unwrap(),
+                )
+                .expect("order key column must be non-NULL")
+                .into_scalar_impl();
+            let encoded_pk = memcmp_encoding::encode_row(
+                (&row).project(
+                    &this
+                        .input_pk_indices
+                        .iter()
+                        .map(|idx| this.col_mapping.upstream_to_state_table(*idx).unwrap())
+                        .collect_vec(),
+                ),
+                &vec![OrderType::ascending(); this.input_pk_indices.len()],
+            )?
+            .into_boxed_slice();
+            let key = StateKey {
+                order_key,
+                encoded_pk,
+            };
+            for (call, state) in this.calls.iter().zip_eq_fast(&mut partition.states) {
+                state.append(
+                    key.clone(),
+                    (&row)
+                        .project(
+                            &call
+                                .args
+                                .val_indices()
+                                .iter()
+                                .map(|idx| this.col_mapping.upstream_to_state_table(*idx).unwrap())
+                                .collect_vec(),
+                        )
+                        .into_owned_row()
+                        .into_inner()
+                        .into(),
+                );
+            }
+        }
+
+        // Ensure states correctness.
+        assert!(partition.is_aligned());
+
+        // Ignore ready windows (all ready windows were outputted before).
+        while partition
+            .states
+            .iter()
+            .all(|state| state.curr_window().is_ready)
+        {
+            partition.states.iter_mut().for_each(|state| {
+                state.slide();
+            });
```

Comment:
IIUC, `slide` here will trigger computation. Though it's lightweight for `LEAD` and `LAG`, is it possible to skip computation for these false readies in case of aggregations?

Reply:
We can't easily distinguish between the first several rows in the stream and first several rows in the buffer state table. A solution will be adding a `outputted` column to the state table, but I'm not sure whether it's worth to do so.

Reply:
I mean avoid computation on recovery. 🥺